pystiche/papers/sanakoyeu_et_al_2018/core.py

```diff
@@ -0,0 +1,302 @@
+import itertools
+from typing import Callable, Optional, Union
+
+import torch
+from torch import nn
+from torch.optim.lr_scheduler import ExponentialLR
+from torch.optim.lr_scheduler import _LRScheduler as LRScheduler
+from torch.optim.optimizer import Optimizer
+from torch.utils.data import DataLoader
+
+from pystiche.loss.perceptual import PerceptualLoss
+from pystiche.ops.container import MultiLayerEncodingOperator
+
+from .data import (
+    sanakoyeu_et_al_2018_dataset,
+    sanakoyeu_et_al_2018_image_loader,
+    sanakoyeu_et_al_2018_images,
+)
+from .loss import (
+    MultiLayerDicriminatorEncodingOperator,
+    SanakoyeuEtAl2018DiscriminatorLoss,
+    sanakoyeu_et_al_2018_discriminator_operator,
+    sanakoyeu_et_al_2018_transformer_loss,
+)
+from .modules import SanakoyeuEtAl2018Discriminator, SanakoyeuEtAl2018Transformer
+from .utils import (
+    ExponentialMovingAverage,
+    sanakoyeu_et_al_2018_lr_scheduler,
+    sanakoyeu_et_al_2018_optimizer,
+)
+
+__all__ = [
+    "sanakoyeu_et_al_2018_training",
+    "sanakoyeu_et_al_2018_stylization",
+    "SanakoyeuEtAl2018Transformer",
+    "SanakoyeuEtAl2018Discriminator",
+    "sanakoyeu_et_al_2018_transformer_loss",
+    "SanakoyeuEtAl2018DiscriminatorLoss",
+    "sanakoyeu_et_al_2018_discriminator_operator",
+    "sanakoyeu_et_al_2018_images",
+    "sanakoyeu_et_al_2018_image_loader",
+    "sanakoyeu_et_al_2018_dataset",
+]
+
+
+def gan_optim_loop(
+    content_image_loader: DataLoader,
+    style_image_loader: DataLoader,
+    transformer: nn.Module,
+    discriminator_operator: MultiLayerEncodingOperator,
+    discriminator_criterion: nn.Module,
+    transformer_criterion: nn.Module,
+    transformer_criterion_update_fn: Callable[
+        [nn.Module, nn.Module, torch.tensor, nn.ModuleDict], None
+    ],
+    discriminator_optimizer: Optional[Optimizer] = None,
+    transformer_optimizer: Optional[Optimizer] = None,
+    get_optimizer: Optional[Callable[[nn.Module], Optimizer]] = None,
+    target_win_rate: float = 0.8,
+    discriminator_success: Optional[ExponentialMovingAverage] = None,
```

Comment:
In case of the `epoch_gan_optim_loop` the `discriminator_success` should not be reset after every epoch, but should simply continue to run.

Reply:
So? Why would a user pass the `ExponentialMovingAverage` in the first place? Can't we just simply create it in this function?

Reply:
Of course, the `dicriminator_success` can only be created here, taking into account that it will not be reset on a second call, as is the case with the `epoch_gan_optim_loop` function. But this would result in a loss of flexibility, because the user can no longer set the parameters of the `ExponentialMovingAverage` differently. Such as a different `init_val` or a different `smoothing_factor` (previously called `alpha`). 

Or am I misinterpreting your thoughts right now?

Reply:
Does he have to? If I remember correctly, the `init_val` should simply be the `target_win_rate`, or am I mistaken? I think we have enough (for too many for most metrics) parameters as is. If the author does not vary this in an experiment we want to replicate, I think we should just create it here and be done with it.